src/main/java/com/hivemq/persistence/local/memory/ClientSessionMemoryLocalPersistence.java

```diff
@@ -0,0 +1,531 @@
+        if (storedSession == null) {
+            return null;
+        }
+
+        final ClientSession clientSession = storedSession.getObject().deepCopyWithoutPayload();
+
+        if (checkExpired &&
+                ClientSessions.isExpired(clientSession, System.currentTimeMillis() - storedSession.getTimestamp())) {
+            return null;
+        }
+        if (includeWill) {
+            dereferenceWillPayload(clientSession);
+        }
+        return clientSession;
+    }
+
+    @Override
+    public @Nullable Long getTimestamp(final @NotNull String clientId) {
+        final int bucketIndex = BucketUtils.getBucket(clientId, bucketCount);
+        return getTimestamp(clientId, bucketIndex);
+    }
+
+    @Override
+    public @Nullable Long getTimestamp(final @NotNull String clientId, final int bucketIndex) {
+
+        final Map<String, PersistenceEntry<ClientSession>> bucket = getBucket(bucketIndex);
+
+        final PersistenceEntry<ClientSession> storedSession = bucket.get(clientId);
+        if (storedSession != null) {
+            return storedSession.getTimestamp();
+        }
+        return null;
+    }
+
+    @Override
+    @ExecuteInSingleWriter
+    public void put(
+            final @NotNull String clientId,
+            final @NotNull ClientSession clientSession,
+            final long timestamp,
+            final int bucketIndex) {
+        checkNotNull(clientId, "Client id must not be null");
+        checkNotNull(clientSession, "Client session must not be null");
+        checkArgument(timestamp > 0, "Timestamp must be greater than 0");
+        ThreadPreConditions.startsWith(SINGLE_WRITER_THREAD_PREFIX);
+
+
+        final Map<String, PersistenceEntry<ClientSession>> sessions = getBucket(bucketIndex);
+        final ClientSession usedSession = clientSession.deepCopyWithoutPayload();
+        final boolean isPersistent = persistent(usedSession);
+
+        sessions.compute(clientId, (ignored, storedSession) -> {
+
+            if (storedSession != null) {
+                final ClientSession oldSession = storedSession.getObject();
+
+                currentMemorySize.addAndGet(-storedSession.getEstimatedSize());
+
+                if (oldSession.getWillPublish() != null) {
+                    removeWillReference(oldSession);
+                }
+
+                final boolean prevIsPersistent = persistent(oldSession);
+
+                if ((isPersistent || usedSession.isConnected()) && (!prevIsPersistent && !oldSession.isConnected())) {
+                    sessionsCount.incrementAndGet();
+                } else if ((prevIsPersistent || oldSession.isConnected()) &&
+                        (!isPersistent && !usedSession.isConnected())) {
+                    sessionsCount.decrementAndGet();
+                }
+
+            } else if (isPersistent || usedSession.isConnected()) {
+                sessionsCount.incrementAndGet();
+            }
```

Comment:
```suggestion
                if (prevIsPersistent || oldSession.isConnected()) {
                    sessionsCount.decrementAndGet();
                }
            }
            if (isPersistent || usedSession.isConnected()) {
                sessionsCount.incrementAndGet();
            }
```

Reply:
usedSession.isConnected() is always true

Reply:
@Florian-Limpoeck is session really always connected in put, do we never put a disconnected session?

Reply:
yes we never put disconnected sessions.